lib/matplotlib/figure.py

```diff
@@ -2308,20 +2329,72 @@ def __init__(self,
 
         self.subplotpars = subplotpars
 
-        # constrained_layout:
-        self._constrained = False
-
-        self.set_tight_layout(tight_layout)
-
         self._axstack = _AxesStack()  # track all figure axes and current axes
         self.clf()
         self._cachedRenderer = None
 
-        self.set_constrained_layout(constrained_layout)
-
         # list of child gridspecs for this figure
         self._gridspecs = []
 
+    def _check_layout_engines_compat(self, old, new):
+        """
+        Helper for set_layout engine
+
+        If the figure has used the old engine and added a colorbar then the
+        value of colorbar_gridspec must be the same on the new engine.
+        """
+        if old is None or old.colorbar_gridspec == new.colorbar_gridspec:
+            return True
+        # colorbar layout different, so check if any colorbars are on the
+        # figure...
+        for ax in self.axes:
+            if hasattr(ax, '_colorbar'):
+                # colorbars list themselvs as a colorbar.
+                return False
+        return True
+
+    def set_layout_engine(self, layout=None, **kwargs):
+        """
+        Set the layout engine for this figure.
+
+        Parameters
+        ----------
+        layout: {'constrained', 'tight'} or `~.LayoutEngine`
+            'constrained' will use `~.ConstrainedLayoutEngine`, 'tight' will
+            use `~.TightLayoutEngine`.  Users and libraries can define their
+            own layout engines as well.
+        kwargs: dict
+            The keyword arguments are passed to the layout engine to set things
+            like padding and margin sizes.  Only used if *layout* is a string.
+        """
+        if layout is None:
+            if mpl.rcParams['figure.autolayout']:
+                layout = 'tight'
+            elif mpl.rcParams['figure.constrained_layout.use']:
+                layout = 'constrained'
+            else:
+                self._layout_engine = None
+                return
+        if layout == 'tight':
+            new_layout_engine = TightLayoutEngine(**kwargs)
+        elif layout == 'constrained':
+            new_layout_engine = ConstrainedLayoutEngine(**kwargs)
+        elif isinstance(layout, LayoutEngine):
+            new_layout_engine = layout
+        else:
+            raise ValueError(f"Invalid value for 'layout': {layout!r}")
+
+        if self._check_layout_engines_compat(self._layout_engine,
```

Comment:
We should either always change it and warn or not change it and raise.  

Not changing and warning makes sense if someone is using us directly and interactively, but if we end up behind any layers of indirection will make it both confusing for users and hard for people wrapping us to program against.

Reply:
I don't think we should change if there is a colorbar; the layout will just break.  

Im not sure I follow how warning and not doing anything is bad.  I mean I guess if someone wraps us and tries to magically switch layout engines behind the users back, that would be a problem, but they shouldn't do that, or if they do, they can add their own more stringent checking.  I think its a better user experience from our end just to warn and not do anything.

Reply:
I still lean towards raising.  If we raise something like a `RuntimeError`, then someone who wants to try and carry on if it fails can do `try: ... except`, however if we only warn someone who wants to know if it has failed has no good way to do so (yes, you can use various tricks to catch warnings, but that involves some global state (under the hood)).

I think this is a case where our dual application vs library identity is causing problems.  In application-mode warning is 100% the right thing to do, in library-mode raising is 100% the right thing to do.

Not going to block over this.

Reply:
This now raises...